### PR TITLE
Implement cyclic ID using IDCD

### PR DESCRIPTION
**Pull request by emilydogbatsePNNL2025**
Part of https://github.com/y0-causal-inference/y0/issues/71

This PR is about implementing the full main cyclic ID algorithm as a top level function that can be called and used. In order to follow TDD I've started implementing the unit tests and integration tests before fully writing the full implementation. 

### Changes Made and files to look at
- Added `id_cyclic.py` for where the code implementation will go.
- Added `initialize_district_distribution()` for a helper (implements Proposition 9.8 in the paper)
- Started a unit tests suite (work in progress on adding this) in the file `test_cyclicID.py`
- I also did a test to see if mathematically the implementation with identify_outcomes was correct in the file called `test_idcd_math.py` (all tests pass)
- Added a draft end-to-end full high level notebook with examples to test the implementation.

### Status so far:
This is a draft and I do plan to finish the tests first. 

### Completed So far:
- Core cyclic ID implementation (**NOTE**: If you see the notebook titled: `Cyclic_ID_High_level.ipynb` you can see the notebook has the full implementation of the main cyclic ID algorithm;  solely for an end-to-end test with IDCD done now. However, still needs tests! 
- Helper function added due to a bug where the interventions were not appearing in results. 
- Manual testing with graph structures
- Identifiable/Unidentifiable test cases with various queries make sense mathematically.

### In Progress:

- Complete the all tests first
- Edge case handling
- Documentation improvements
- Finish with algorithm implementation

### CODE REVIEW

**Comment by cthoyt** [comment]
i made a few minor tweaks - I'd say directly go into the python module and take it all out of the notebook as soon as possible

**Comment by emilydogbatsePNNL2025** [comment]
@cthoyt thanks for the fixes. working on that now!

File: notebooks/Cyclic_ID_High_level.ipynb
Comment (cthoyt) [inline notebooks/Cyclic_ID_High_level.ipynb]: do we need this anymore?

**Review by cthoyt — COMMENTED** [review]
@emilydogbatsePNNL2025 this is a great next step! you can take a look at some of the minor tweaks I added. We also need explicit expressions for all of the tests

File: src/y0/algorithm/identify/cyclic_id.py [hunk 1]
```diff
@@ -361,10 +455,47 @@ def get_apt_order_predecessors(
...
+    # find all SCCs (feedback loops) within the district
+    district_subgraph = graph.subgraph(district)
+    sccs = get_strongly_connected_components(district_subgraph)
+
+    loop_distributions = []
+
+    for scc in sccs:
+        predecessors = _get_predecessors(scc, apt_order)
+        # compute P(SCC | predecessors) or P(SCC) if no predecessors
+        distribution = Probability.safe(scc.union(predecessors)).conditional(predecessors)
```
[the remark this answers is not in the document]
Comment (cthoyt) [reply src/y0/algorithm/identify/cyclic_id.py]: I externalized this in its own function in b09c4dc

File: notebooks/Cyclic_ID_High_level.ipynb
Comment (emilydogbatsePNNL2025) [reply notebooks/Cyclic_ID_High_level.ipynb]: Yes @djinnome mentioned he wanted a high level tutorial notebook showing how to use it once everything is implemented fully with a couple examples.

**Comment by emilydogbatsePNNL2025** [comment]
@cthoyt thanks for the fixes! I'll take a look at them and then add the explicit expressions for the tests

**Comment by cthoyt** [comment]
@emilydogbatsePNNL2025 I added some improvements to product simplification in #359 (so you can add `.simplify()` at the end to resolve fractions

reminder that sorting is built in to `Product.safe` too

File: notebooks/Cyclic_ID_High_level.ipynb
Comment (emilydogbatsePNNL2025) [reply notebooks/Cyclic_ID_High_level.ipynb]: that makes sense! I can remove it while finishing up these last few fixes on this PR.
